cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,523 @@
+		data.removeNode(node.Node().Name)
+	}
+
+	return data.addNodeInfo(node)
+}
+
+func (data *internalDeltaSnapshotData) removeNode(nodeName string) error {
+	_, found := data.nodeInfoMap[nodeName]
+	if found {
+		// If node was added or modified within this delta, delete this change.
+		delete(data.nodeInfoMap, nodeName)
+	}
+
+	ni, err := data.baseData.GetNodeInfo(nodeName)
+	if err == nil && ni != nil {
+		// If node was found in the underlying data, mark it as deleted in delta.
+		data.deletedNodeInfos[nodeName] = true
+	}
+
+	if err != nil && !found {
+		// Node not found in the chain.
+		return err
+	}
+
+	// Maybe consider deleting from the lists instead. Maybe not.
+	data.clearCaches()
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addPod(pod *apiv1.Pod, nodeName string) error {
+	if _, found := data.nodeInfoMap[nodeName]; !found {
+		if ni, err := data.baseData.GetNodeInfo(nodeName); err != nil {
+			return err
+		} else {
+			data.nodeInfoMap[nodeName] = ni.Clone()
+		}
+	}
+
+	data.nodeInfoMap[nodeName].AddPod(pod)
+	if data.podList != nil || data.havePodsWithAffinity != nil {
+		data.clearPodCaches()
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) getNodeForPod(namespace, name string) (string, error) {
+	if data == nil {
+		return "", fmt.Errorf("pod %s/%s not in snapshot", namespace, name)
+	}
+
+	for nodeName, nodeInfo := range data.nodeInfoMap {
+		for _, pod := range nodeInfo.Pods() {
+			if pod.Name == name && pod.Namespace == namespace {
+				return nodeName, nil
+			}
+		}
+	}
+
+	return data.baseData.getNodeForPod(namespace, name)
+}
+
+func (data *internalDeltaSnapshotData) removePod(namespace string, name string) error {
+	nodeName, err := data.getNodeForPod(namespace, name)
+	if err != nil {
+		return err
+	}
+
+	if _, found := data.nodeInfoMap[nodeName]; !found {
+		if ni, err := data.baseData.GetNodeInfo(nodeName); err != nil {
+			return err
+		} else {
+			data.nodeInfoMap[nodeName] = ni.Clone()
+		}
+	}
+
+	nodeInfo, found := data.nodeInfoMap[nodeName]
+	if !found {
+		return fmt.Errorf("internal error: node not found")
+	}
+
+	preAffinityPods := len(nodeInfo.PodsWithAffinity())
+	for _, pod := range nodeInfo.Pods() {
+		if pod.Namespace == namespace && pod.Name == name {
+			if err := nodeInfo.RemovePod(pod); err != nil {
+				return fmt.Errorf("cannot remove pod; %v", err)
+			}
+			break
+		}
+	}
+
+	// Maybe consider deleting from the list in the future. Maybe not.
+	postAffinityPods := len(nodeInfo.PodsWithAffinity())
+	if preAffinityPods == 1 && postAffinityPods == 0 {
+		data.havePodsWithAffinity = nil
+	}
+	if data.podList != nil || data.rawPods != nil {
+		data.clearPodCaches()
+	}
+
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) getRawPods() ([][]*apiv1.Pod, int) {
+	if data == nil {
+		return [][]*apiv1.Pod{}, 0
+	}
+	if data.rawPods != nil {
+		return data.rawPods, data.rawPodsTotal
+	}
+	basePods, baseCount := data.baseData.getRawPods()
```

Comment:
Assuming we keep the method name `getRawPods` add the `raw` into variable name.
So `baseRawPods`.
Also use `basePodsTotal` or `baseRawPodsTotal`

Reply:
renamed